scheduler/src/cook/authorization.clj

```diff
@@ -0,0 +1,180 @@
+  `:authorization-fn` config file key to it. 
+
+  For example, you could consult an external authorization system that
+  manages user credits or billing, or you could implement a \"teams\"
+  feature, where users are allowed to modify their teammates' jobs.
+
+"
+  (:require [cook.global-state :refer [global-state]]
+            [cook.util :refer [lazy-load-var]]
+            [clojure.tools.logging :as log]
+            [plumbing.core :refer (fnk defnk)]))
+
+
+;; The defonce is a kludge to work around
+;; https://groups.google.com/forum/#!topic/clojure/SYYYwZIrFiY -
+;; reloading a protocol causes existing objects of that type to become
+;; invalid as instances of the new protocol.
+;;
+;; If you redefine this protocol at runtime, you'll have to manually
+;; remove the defonce. Other than that, this should not cause issues.
+(defonce __ownable
+  (defprotocol Ownable
+    "Functions to be defined on types that are able to be owned by a specific user."
+    (owner [this] "Returns the username that owns this object.")))
+
+(def system (reify Ownable (owner [this] ::system)))
+
+;;
+;; Authorization functions
+;; 
+
+
+(defn open-auth
+  "This authorization function allows any user to do anything to any
+  object. It is intended for testing and development."
+  [settings user verb object]
+  true)
+
+
+(defn configfile-admins-auth
+  "This authorization function consults the set of usernames specified
+  in the :admins key of the config file.
+
+  Usernames in this set are administrators, who are allowed to do anything to any object.
+  Non-admins are only allowed to manipulate objects that they own."
+  [settings
+   ^String user
+   ^clojure.lang.Keyword verb
+   ^cook.authorization.Ownable object ]
+  (let [admins    (:admins settings)
+        is-admin? (contains? admins user)
+        owner     (owner object)]
+    (log/debug "[configfile-admins-auth] Checking whether user" user
+               "may perform" verb
+               "on object" (str object) "."
+               "Admins are:" admins)
+    (cond is-admin? (do
+                      (log/debug "[configfile-admins-auth] User" user "is an admin, allowing.")
+                      true)
+          (= owner user) (do
+                           (log/debug "[configfile-admins-auth] Object is owned by user, allowing.")
+                           true)
+          :else (do 
+                  (log/info "[configfile-admins-auth] Unauthorized access attempt: user" user
+                            "is not allowed to perform" verb "on" (str object ",") "denying.")
+                  false))))
+
+
+(defn is-authorized?
+  "Determines whether the given user can perform the given operation
+  on the given object, according to the authorization function
+  specified in the config file.
+
+  Returns true if allowed, else false."
+  ([^String user
+     ^clojure.lang.Keyword verb
+     ^cook.authorization.Ownable object] 
+     (is-authorized? (-> @global-state :settings) user verb object))
```

Comment:
I'm not a fan of global state either, but it's already deeply embedded in the app's design. The app uses a bunch of Java libraries to interact with Mesos and the web, which are very much stateful and in the object-oriented paradigm. The global state (among other things) keeps track of those Java objects. 

Short of redesigning the entire app, there is not much that can be done now except fence the global state off into a designated "stateful" containment area so we can keep it from expanding and at least know what's in it. Over time, we can work on whittling that down -- I'm sure a lot of it can be removed and converted into non-global state and/or pure functions. For now (this version), I'm aiming at just organizing and fencing off what's already there, since a rewrite of all that was not in the remit.

Speaking more generally, all long-lived daemon apps have to have *some* kind of global state, if only a network socket or a file handle (or else they wouldn't be able to communicate with the outside world at all.) I think the best we can hope for is to isolate and manage that. 99% of the code can be purely functional and take state as an argument, but trying to carry that concept through to its logical conclusion leads to a different problem: all functions grow extra arguments for passing around database handles or things like that, which then have to be threaded through the entire call stack every single time. Adding a new item then means reworking every function in the call stack to thread that data through.

Clojure being a pragmatic language, we have the STM available for just such cases.